VSR: Panic on checkpoint divergence

sentientwaffle opened the pull request:
### Panic on checkpoint divergence

In the original state sync design, checkpoint divergence (e.g. due to storage engine nondeterminism) was tolerated – within a given bound, the cluster could use state sync to automatically recover. The rationale for automatically recovering from checkpoint divergence was that it would simplify replica version upgrades. In particular, it was to add some guardrails case an upgrade that was not expected to change determinism did change it.

However, after some discussion we decided that storage engine nondeterminism is Not Allowed. (Specifically, mixed-version clusters will be forbidden.) If divergence is detected, we will now abort, to escalate to manual operator intervention.

### Assume all checkpoints are canonical

Assume that all checkpoints are canonical checkpoints.

In the original design of state sync, the distinction of canonical/uncanonical checkpoints enabled the cluster to (automatically) recover from a divergent checkpoint.
Since checkpoint divergence is now forbidden (on penalty of process-abort) this distinction is no longer necessary.

This fixes the liveness issue formerly guarded by `Simulator.core_missing_checkpoint()`, and avoids a crash in that same function:

- Fixes https://github.com/tigerbeetle/tigerbeetle/issues/1382
- Fixes https://github.com/tigerbeetle/tigerbeetle/issues/1430

The new logic to answer "Can replica X sync to checkpoint Y?" is:
1. the new checkpoint must advance the replica's op, and
2. the new checkpoint must either be committed atop _or_ more than one checkpoint ahead of our current checkpoint.
    - If a checkpoint is committed atop, that guarantees that a (commit) quorum of replicas has reached that checkpoint.
    - See `Replica.jump_sync_target()` for more details about syncing to an idle cluster.

### Remove `VSRState.commit_min_canonical`

`commit_min_canonical` was only necessary because replicas needed to (sometimes) re-sync a checkpoint when recovering from a divergent checkpoint.

sentientwaffle committed:
VSR/Sync: Panic on checkpoint divergence

In the original state sync design, checkpoint divergence (e.g. due to storage engine nondeterminism) was tolerated – within a given bound, the cluster could use state sync to automatically recover. The rationale for automatically recovering from checkpoint divergence was that it would simplify replica version upgrades. In particular, it was to add some guardrails case an upgrade that was not expected to change determinism did change it.

However, after some discussion we decided that storage engine nondeterminism is Not Allowed. (Specifically, mixed-version clusters will be forbidden.) If divergence is detected, we will now abort, to escalate to manual operator intervention.

sentientwaffle committed:
VSR: Assume all checkpoints are canonical

Assume that all checkpoints are canonical checkpoints.

In the original design of state sync, the distinction of canonical/uncanonical checkpoints enabled the cluster to (automatically) recover from a divergent checkpoint.
Since checkpoint divergence is now forbidden (on penalty of process-abort) this distinction is no longer necessary.

This fixes the liveness issue formerly guarded by `Simulator.core_missing_checkpoint()`, and avoids a crash in that same function:

- Fixes https://github.com/tigerbeetle/tigerbeetle/issues/1382
- Fixes https://github.com/tigerbeetle/tigerbeetle/issues/1430

The new logic to answer "Can replica X sync to checkpoint Y?" is:
1. the new checkpoint must advance the replica's op, and
2. the new checkpoint must either be committed atop _or_ more than one checkpoint ahead of our current checkpoint.
    - If a checkpoint is committed atop, that guarantees that a (commit) quorum of replicas has reached that checkpoint.
    - See `Replica.jump_sync_target()` for more details about syncing to an idle cluster.

sentientwaffle committed:
Simulator: Remove core_missing_checkpoint

sentientwaffle committed:
VSR: Remove VSRState.commit_min_canonical

`commit_min_canonical` was only necessary because replicas needed to (sometimes) re-sync a checkpoint when recovering from a divergent checkpoint.

sentientwaffle commented on src/testing/cluster/sync_checker.zig:
I'm thinking about just removing this checker now, since we panic on divergent checkpoints anyway.

matklad replied:
Yeah, I it would be enough to just assert that the final checkpoint is the same after we call `assert_cluster_convergence()`.

That is, I thin relying on _just_ replicas internally panicking on divergence is not enough  (what if that code is buggy), but just sanity checking it _once_ in the end should be good!

sentientwaffle replied:
Very good point. Rewritten to:

> Every sync target **must**:
> - have an op greater than the syncing replica's current checkpoint op.
> - either:
>   - be committed atop – i.e. the syncing replica can sync to `healthy_replica.checkpoint_op` when `trigger_for_checkpoint(healthy_replica.checkpoint_op) < healthy_replica.commit_min` – ensuring that the checkpoint has been reached by a quorum of replicas, or
>   - be more than 1 checkpoint ahead of our current checkpoint.

matklad commented on docs/internals/sync.md:
```diff
@@ -143,23 +142,17 @@ A checkpoint identifier is attached to the following message types:
+Every sync target **must**:
+- have an op greater than the replica's current checkpoint op.
+- either:
+  - be committed atop (ensuring that it has been reached by a quorum of replicas), or
```
Could you expand what "committed atop" means? To disambiguate between `op_checkpoint + 1` and
`Checkpoint.trigger_for(op_checkpoint) + 1`, which are both valid interpretations for the current
wording.

matklad commented on src/vsr/replica.zig:
```diff
@@ -8357,30 +8353,58 @@ pub fn ReplicaType(
+            // If R₁/R₂ state-sync to R₀'s latest checkpoint, then a single block corruption on R₀
+            // could lead to permanent unavailability.
+            if (candidate.checkpoint_op == self.op_checkpoint_next()) {
+                if (header.into_const(.commit)) |h| {
+                    if (h.commit <= self.op_checkpoint_next_trigger()) return;
+                } else {
+                    return;
+                }
```
:thinking: it feels like we might be _not_ jumping checkpoint in some cases here. 

In particular, if a cluster is stuck in a view change, then no commit messages are flowing, and so a replica can't learn that a checkpoint that it gets from pings is commited on top of. 

Though, this is _probably_ fine --- the fact that replica lags by a checkpoint doesn't prevent it from participating in a view change, and, once view change happens, `commit` messages start flowing back.

matklad replied:
I guess that's one more argument for merging pings&commits

sentientwaffle replied:
> Though, this is probably fine --- the fact that replica lags by a checkpoint doesn't prevent it from participating in a view change, and, once view change happens, commit messages start flowing back.

Yeah, I spent some time trying to come up with a case where this doesn't work.

The closest I found was this liveness issue, which we work around in the simulator: https://github.com/tigerbeetle/tigerbeetle/pull/933#discussion_r1245440623

> I guess that's one more argument for merging pings&commits

Agreed.

sentientwaffle committed:
Simulator: Check checkpoint_id convergence

sentientwaffle committed:
Testing: Remove SyncChecker

Now that all checkpoints are canonical, SyncChecker doesn't check anything useful.

matklad commented on src/simulator.zig:
```diff
@@ -412,6 +410,8 @@ pub const Simulator = struct {
 
         simulator.cluster.state_checker.assert_cluster_convergence();
 
+        var checkpoint_id: ?u128 = null;
```
This variable seems to be never assigned to

sentientwaffle replied:
Ahh I "refactored" at the last minute :facepalm:

matklad replied:
Hopefully the very next version of zig will catch these things!

sentientwaffle replied:
I reworked this -- the checkpoint id is now tested in its own loop _after_ the "repairing" loop. That means it can `assert` that the checkpoint ids match instead of just `return`ing if they don't.
